javaparser-core/src/main/javacc/java.jj

```diff
@@ -672,6 +672,74 @@ TOKEN :
 /*
  * Program structuring syntax follows.
  */
+IndexUnit IndexUnit():
+{
+	/* Actual Type */
+	NodeList<CompilationUnit> compilationUnits = emptyList();
+	CompilationUnit cu;
+
+	/* Parsing the compilation unit itself */
+	PackageDeclaration pakage = null;
+    NodeList<ImportDeclaration> imports = emptyList();
+    ImportDeclaration in = null;
+    NodeList<TypeDeclaration<?>> types = emptyList();
+    ModifierHolder modifier;
+    TypeDeclaration tn = null;
+    ModuleDeclaration module = null;
+}
+{
+	( LOOKAHEAD(2) ";" )*
+    [ LOOKAHEAD(PackageDeclaration()) pakage = PackageDeclaration() ]
+    (
+        in = ImportDeclaration() { imports = add(imports, in); }
+        |
+        (
+            modifier = Modifiers()
+            (
+                tn = ClassOrInterfaceDeclaration(modifier) { types = add(types, tn); }
+            |
+                tn = EnumDeclaration(modifier) { types = add(types, tn); }
+            |
+                tn = AnnotationTypeDeclaration(modifier) { types = add(types, tn); }
+            |
+                module = ModuleDeclaration(modifier)
+            |
+                ";"
+            )
+        )
+        {
+             cu = new CompilationUnit(range(token_source.getHomeToken(), token()), pakage, imports, types, module);
+             compilationUnits.add(cu);
+        }
+    )
+
+    [ LOOKAHEAD(PackageDeclaration()) pakage = PackageDeclaration() ]
+        (
+            in = ImportDeclaration() { imports = add(imports, in); }
+            |
+            (
+                modifier = Modifiers()
+                (
+                    tn = ClassOrInterfaceDeclaration(modifier) { types = add(types, tn); }
+                |
+                    tn = EnumDeclaration(modifier) { types = add(types, tn); }
+                |
+                    tn = AnnotationTypeDeclaration(modifier) { types = add(types, tn); }
+                |
+                    module = ModuleDeclaration(modifier)
+                |
+                    ";"
+                )
+            )
+            {
+                 cu = new CompilationUnit(range(token_source.getHomeToken(), token()), pakage, imports, types, module);
```

Comment:
This seems to require a correct, complete compilation unit.  Is it your plan in the future to relax this restriction so that various parts (such as method bodies) can be omitted?  That's fine, but the scope of this particular pull request wasn't clear from the description.